File: packages/infinity_context_sdk/infinity_context_sdk/anchors.py
"""Semantic anchor helpers for the public Infinity Context SDK."""

from __future__ import annotations

from typing import Any

import infinity_context_sdk._payloads as _payloads


class InfinityContextAnchorsMixin:
    def create_anchor(
        self,
        *,
        kind: str,
        label: str,
        aliases: list[str] | None = None,
        description: str | None = None,
        confidence: str | None = None,
        evidence_refs: list[dict[str, Any]] | None = None,
        observed_at: str | None = None,
        valid_from: str | None = None,
        valid_to: str | None = None,
        metadata: dict[str, Any] | None = None,
        space_id: str | None = None,
        memory_scope_id: str | None = None,
        space_slug: str | None = None,
        memory_scope_external_ref: str | None = None,
    ) -> dict[str, Any]:
        return self._request(
            "POST",
            "/v1/anchors",
            json=_payloads.without_none(
                {
                    **_payloads.single_scope_body(
                        space_id=space_id,
                        memory_scope_id=memory_scope_id,
                        thread_id=None,
                        space_slug=space_slug,
                        memory_scope_external_ref=memory_scope_external_ref,
                        thread_external_ref=None,
                    ),
                    "kind": kind,
                    "label": label,
                    "aliases": aliases or [],
                    "description": description,
                    "confidence": confidence,
                    "evidence_refs": evidence_refs if evidence_refs is not None else None,
                    "observed_at": observed_at,
                    "valid_from": valid_from,
                    "valid_to": valid_to,
                    "metadata": metadata or {},
                }
            ),
        )

    def list_anchors(
        self,
        *,
        space_id: str | None = None,
        memory_scope_id: str | None = None,
        space_slug: str | None = None,
        memory_scope_external_ref: str | None = None,
        kind: str | None = None,
        status: str | None = "active",
        limit: int = 100,
    ) -> dict[str, Any]:
        return self._request(
            "GET",
            "/v1/anchors",
            params=_payloads.without_none(
                {
                    **_payloads.single_scope_body(
                        space_id=space_id,
                        memory_scope_id=memory_scope_id,
                        thread_id=None,
                        space_slug=space_slug,
                        memory_scope_external_ref=memory_scope_external_ref,
                        thread_external_ref=None,
                    ),
                    "kind": kind,
                    "status": status,
                    "limit": limit,
                }
            ),
        )

    def list_anchor_relations(
        self,
        *,
        space_id: str | None = None,
        memory_scope_id: str | None = None,
        space_slug: str | None = None,
        memory_scope_external_ref: str | None = None,
        status: str | None = "active",
        limit: int = 100,
        anchor_limit: int = 500,
    ) -> dict[str, Any]:
        return self._request(
            "GET",
            "/v1/anchors/relations",
            params=_payloads.without_none(
                {
                    **_payloads.single_scope_body(
                        space_id=space_id,
                        memory_scope_id=memory_scope_id,
                        thread_id=None,
                        space_slug=space_slug,
                        memory_scope_external_ref=memory_scope_external_ref,
                        thread_external_ref=None,
                    ),
                    "status": status,
                    "limit": limit,
                    "anchor_limit": anchor_limit,
                }
            ),
        )

    def update_anchor(
        self,
        anchor_id: str,
        *,
        label: str | None = None,
        aliases: list[str] | None = None,
        description: str | None = None,
        confidence: str | None = None,
        evidence_refs: list[dict[str, Any]] | None = None,
        observed_at: str | None = None,
        valid_from: str | None = None,
        valid_to: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> dict[str, Any]:
        return self._request(
            "PATCH",
            f"/v1/anchors/{anchor_id}",
            json=_payloads.without_none(
                {
                    "label": label,
                    "aliases": aliases or [],
                    "description": description,
                    "confidence": confidence,
                    "evidence_refs": evidence_refs if evidence_refs is not None else None,
                    "observed_at": observed_at,
                    "valid_from": valid_from,
                    "valid_to": valid_to,
                    "metadata": metadata or {},
                }
            ),
        )

    def delete_anchor(
        self,
        anchor_id: str,
        *,
        reason: str = "manual delete",
    ) -> dict[str, Any]:
        return self._request(
            "DELETE",
            f"/v1/anchors/{anchor_id}",
            json={"reason": reason},
        )

    def backfill_anchors(
        self,
        *,
        space_id: str | None = None,
        memory_scope_id: str | None = None,
        space_slug: str | None = None,
        memory_scope_external_ref: str | None = None,
        limit_per_source: int = 100,
    ) -> dict[str, Any]:
        return self._request(
            "POST",
            "/v1/anchors/backfill",
            json=_payloads.without_none(
                {
                    **_payloads.single_scope_body(
                        space_id=space_id,
                        memory_scope_id=memory_scope_id,
                        thread_id=None,
                        space_slug=space_slug,
                        memory_scope_external_ref=memory_scope_external_ref,
                        thread_external_ref=None,
                    ),
                    "limit_per_source": limit_per_source,
                }
            ),
        )

    def list_anchor_merge_suggestions(
        self,
        *,
        space_id: str | None = None,
        memory_scope_id: str | None = None,
        space_slug: str | None = None,
        memory_scope_external_ref: str | None = None,
        kind: str | None = None,
        limit: int = 50,
    ) -> dict[str, Any]:
        return self._request(
            "GET",
            "/v1/anchors/merge-suggestions",
            params=_payloads.without_none(
                {
                    **_payloads.single_scope_body(
                        space_id=space_id,
                        memory_scope_id=memory_scope_id,
                        thread_id=None,
                        space_slug=space_slug,
                        memory_scope_external_ref=memory_scope_external_ref,
                        thread_external_ref=None,
                    ),
                    "kind": kind,
                    "limit": limit,
                }
            ),
        )

    def merge_anchor(
        self,
        source_anchor_id: str,
        *,
        target_anchor_id: str,
        reason: str,
    ) -> dict[str, Any]:
        return self._request(
            "POST",
            f"/v1/anchors/{source_anchor_id}/merge",
            json={"target_anchor_id": target_anchor_id, "reason": reason},
        )

    def split_anchor(
        self,
        anchor_id: str,
        *,
        alias: str,
        new_label: str | None = None,
        reason: str = "manual split",
    ) -> dict[str, Any]:
        return self._request(
            "POST",
            f"/v1/anchors/{anchor_id}/split",
            json=_payloads.without_none(
                {
                    "alias": alias,
                    "new_label": new_label,
                    "reason": reason,
                }
            ),
        )

File: packages/infinity_context_ts_sdk/src/resources/anchors.ts
import type { RequestExecutor } from "../client.js";
import { scopeQuery, withoutUndefined, type SingleScopeInput } from "../payload.js";
import type { AnchorRecord, ApiEnvelope, JsonObject, SourceRef } from "../types.js";

type AnchorScopeInput = Omit<SingleScopeInput, "threadId" | "threadExternalRef">;

export interface AnchorBackfillSource extends JsonObject {
  readonly source_type: string;
  readonly scanned: number;
  readonly observed: number;
  readonly skipped_conflicts: number;
}

export interface AnchorBackfillData extends JsonObject {
  readonly anchors: readonly AnchorRecord[];
  readonly created: number;
  readonly updated: number;
  readonly sources: readonly AnchorBackfillSource[];
  readonly diagnostics: JsonObject;
}

export interface AnchorMergeCandidate extends JsonObject {
  readonly source_anchor: AnchorRecord;
  readonly target_anchor: AnchorRecord;
  readonly confidence: string;
  readonly score: number;
  readonly reasons: readonly string[];
  readonly metadata: JsonObject;
}

export class AnchorsClient {
  constructor(private readonly http: RequestExecutor) {}

  createAnchor(input: SingleScopeInput & {
    readonly kind: string;
    readonly label: string;
    readonly aliases?: readonly string[];
    readonly description?: string;
    readonly confidence?: string;
    readonly evidenceRefs?: readonly SourceRef[];
    readonly observedAt?: string;
    readonly validFrom?: string;
    readonly validTo?: string;
    readonly metadata?: JsonObject;
  }): Promise<ApiEnvelope<AnchorRecord>> {
    return this.http.request<ApiEnvelope<AnchorRecord>>({
      method: "POST",
      path: "/v1/anchors",
      json: withoutUndefined({
        ...scopeQuery(input),
        kind: input.kind,
        label: input.label,
        aliases: input.aliases ?? [],
        description: input.description,
        confidence: input.confidence,
        evidence_refs: input.evidenceRefs,
        observed_at: input.observedAt,
        valid_from: input.validFrom,
        valid_to: input.validTo,
        metadata: input.metadata ?? {},
      }) as JsonObject,
    });
  }

  listAnchors(input: SingleScopeInput & {
    readonly kind?: string;
    readonly status?: string | null;
    readonly limit?: number;
  }): Promise<ApiEnvelope<AnchorRecord[]>> {
    return this.http.request<ApiEnvelope<AnchorRecord[]>>({
      method: "GET",
      path: "/v1/anchors",
      params: withoutUndefined({
        ...scopeQuery(input),
        kind: input.kind,
        status: input.status === undefined ? "active" : input.status,
        limit: input.limit ?? 100,
      }),
    });
  }

  listAnchorRelations(input: SingleScopeInput & {
    readonly status?: string | null;
    readonly limit?: number;
    readonly anchorLimit?: number;
  }): Promise<ApiEnvelope<JsonObject[]>> {
    return this.http.request<ApiEnvelope<JsonObject[]>>({
      method: "GET",
      path: "/v1/anchors/relations",
      params: withoutUndefined({
        ...scopeQuery(input),
        status: input.status === undefined ? "active" : input.status,
        limit: input.limit ?? 100,
        anchor_limit: input.anchorLimit ?? 500,
      }),
    });
  }

  updateAnchor(anchorId: string, input: {
    readonly label?: string;
    readonly aliases?: readonly string[];
    readonly description?: string;
    readonly confidence?: string;
    readonly evidenceRefs?: readonly SourceRef[];
    readonly observedAt?: string;
    readonly validFrom?: string;
    readonly validTo?: string;
    readonly metadata?: JsonObject;
  }): Promise<ApiEnvelope<AnchorRecord>> {
    return this.http.request<ApiEnvelope<AnchorRecord>>({
      method: "PATCH",
      path: `/v1/anchors/${anchorId}`,
      json: withoutUndefined({
        label: input.label,
        aliases: input.aliases,
        description: input.description,
        confidence: input.confidence,
        evidence_refs: input.evidenceRefs,
        observed_at: input.observedAt,
        valid_from: input.validFrom,
        valid_to: input.validTo,
        metadata: input.metadata,
      }) as JsonObject,
    });
  }

  deleteAnchor(anchorId: string, input: { readonly reason?: string } = {}): Promise<ApiEnvelope<AnchorRecord>> {
    return this.http.request<ApiEnvelope<AnchorRecord>>({
      method: "DELETE",
      path: `/v1/anchors/${anchorId}`,
      json: { reason: input.reason ?? "manual delete" },
    });
  }

  backfillAnchors(input: AnchorScopeInput & {
    readonly limitPerSource?: number;
  }): Promise<ApiEnvelope<AnchorBackfillData>> {
    return this.http.request<ApiEnvelope<AnchorBackfillData>>({
      method: "POST",
      path: "/v1/anchors/backfill",
      json: withoutUndefined({
        ...scopeQuery(input),
        limit_per_source: input.limitPerSource ?? 100,
      }) as JsonObject,
    });
  }

  listAnchorMergeSuggestions(input: AnchorScopeInput & {
    readonly kind?: string;
    readonly limit?: number;
  }): Promise<ApiEnvelope<AnchorMergeCandidate[]>> {
    return this.http.request<ApiEnvelope<AnchorMergeCandidate[]>>({
      method: "GET",
      path: "/v1/anchors/merge-suggestions",
      params: withoutUndefined({
        ...scopeQuery(input),
        kind: input.kind,
        limit: input.limit ?? 50,
      }),
    });
  }

  mergeAnchor(
    sourceAnchorId: string,
    input: { readonly targetAnchorId: string; readonly reason: string },
  ): Promise<ApiEnvelope<AnchorRecord>> {
    return this.http.request<ApiEnvelope<AnchorRecord>>({
      method: "POST",
      path: `/v1/anchors/${sourceAnchorId}/merge`,
      json: {
        target_anchor_id: input.targetAnchorId,
        reason: input.reason,
      },
    });
  }

  splitAnchor(
    anchorId: string,
    input: { readonly alias: string; readonly newLabel?: string; readonly reason?: string },
  ): Promise<ApiEnvelope<AnchorRecord>> {
    return this.http.request<ApiEnvelope<AnchorRecord>>({
      method: "POST",
      path: `/v1/anchors/${anchorId}/split`,
      json: withoutUndefined({
        alias: input.alias,
        new_label: input.newLabel,
        reason: input.reason ?? "manual split",
      }) as JsonObject,
    });
  }
}
